### ISSUE TITLE
podman build inside a container is slow

### ISSUE THREAD

**Issue by LHCGreg** [containers/buildah#3436]
(Creating the issue here as directed by the podman issue template for issues with podman build)

**Description**

Even simple build steps like `RUN mkdir /dir1` in a `podman build` inside a container are taking several seconds each. Having a multistage build seems to make it worse?

**Steps to reproduce the issue:**
1. Create an image with podman installed. Here is a minimal, stripped down version of the dockerfile for the *image to run podman in* (not the image to ultimately build with podman). You could build this with podman but I used `docker build . -f perf_test.Dockerfile -t podman_perf_test`

**perf_test.Dockerfile:**
```
FROM python:3.9-slim-buster AS common

FROM common AS builder

RUN \
apt-get update && \
apt-get -y --no-install-recommends install wget gnupg

# No need to bother with cleaning up in the builder step

RUN wget -O kubic_release.key https://download.opensuse.org/repositories/devel:/kubic:/libcontainers:/testing/Debian_Unstable/Release.key && gpg -o /usr/share/keyrings/kubic.gpg --dearmor kubic_release.key

# Debian buster's slirp4netns is 0.2.3, too old. Using bullseye's package might be possible, but mixing packages from different releases can cause problems.
# slirp4netns offers statically linked binaries.
RUN wget -O /usr/local/bin/slirp4netns "https://github.com/rootless-containers/slirp4netns/releases/download/v1.1.11/slirp4netns-$(uname -m)" && chmod +x /usr/local/bin/slirp4netns

FROM common AS final

COPY --from=builder /usr/share/keyrings/kubic.gpg /usr/share/keyrings/kubic.gpg

RUN \

# Set up repository for podman
# Get from repo for debian unstable because currently only that has the latest podman
echo 'deb [signed-by=/usr/share/keyrings/kubic.gpg] http://download.opensuse.org/repositories/devel:/kubic:/libcontainers:/testing/Debian_Unstable/ /' > /etc/apt/sources.list.d/kubic_libcontainers_debian_unstable.list && \

# Add buster backports repository for newer libseccomp2 that podman needs
echo 'deb http://deb.debian.org/debian buster-backports main' > /etc/apt/sources.list.d/buster_backports.list && \

# containernetworking-plugins: Needed by podman?
# fuse-overlayfs: Needed by podman
# libseccomp2: Needed by podman. Need buster-backports for newer version.
# podman: Used for building and inspecting Docker images.
# tini: Used for entrypoint to handle signals correctly
# uidmap: Needed by podman

# XXX: Installing podman installs:
# python3 python3-minimal python3-pkg-resources python3-protobuf python3-six python3.7 python3.7-minimal
# Installing distro python in the python docker images creates potential for conflict

apt-get update && \
apt-get -y --no-install-recommends -t buster-backports install libseccomp2 && \
apt-get -y --no-install-recommends install containernetworking-plugins fuse-overlayfs podman tini uidmap && \
apt-get clean && \
rm -rf /tmp/* /var/tmp/* && \
rm -rf /var/lib/apt/lists/* && \
rm -rf /root/.cache

COPY --from=builder /usr/local/bin/slirp4netns /usr/local/bin/slirp4netns

COPY --chown=root:root docker_files/storage.conf /etc/containers/storage.conf
COPY --chown=root:root docker_files/containers.conf /usr/share/containers/containers.conf


ENTRYPOINT ["/usr/bin/tini", "-g", "--"]
```

*docker_files/storage.conf:*

```
# This file is is the configuration file for all tools
# that use the containers/storage library.
# See man 5 containers-storage.conf for more information
# The "container storage" table contains all of the server options.
[storage]

# Default Storage Driver, Must be set for proper operation.
driver = "overlay"

# Temporary storage location
runroot = "/run/containers/storage"

# Primary Read/Write location of container storage
graphroot = "/var/lib/containers/storage"

# Storage path for rootless users
#
# rootless_storage_path = "$HOME/.local/share/containers/storage"

[storage.options]
# Storage options to be passed to underlying storage drivers

# AdditionalImageStores is used to pass paths to additional Read/Only image stores
# Must be comma separated list.
additionalimagestores = [
]

# Remap-UIDs/GIDs is the mapping from UIDs/GIDs as they should appear inside of
# a container, to the UIDs/GIDs as they should appear outside of the container,
# and the length of the range of UIDs/GIDs.  Additional mapped sets can be
# listed and will be heeded by libraries, but there are limits to the number of
# mappings which the kernel will allow when you later attempt to run a
# container.
#
# remap-uids = 0:1668442479:65536
# remap-gids = 0:1668442479:65536

# Remap-User/Group is a user name which can be used to look up one or more UID/GID
# ranges in the /etc/subuid or /etc/subgid file.  Mappings are set up starting
# with an in-container ID of 0 and then a host-level ID taken from the lowest
# range that matches the specified name, and using the length of that range.
# Additional ranges are then assigned, using the ranges which specify the
# lowest host-level IDs first, to the lowest not-yet-mapped in-container ID,
# until all of the entries have been used for maps.
#
# remap-user = "containers"
# remap-group = "containers"

# Root-auto-userns-user is a user name which can be used to look up one or more UID/GID
# ranges in the /etc/subuid and /etc/subgid file.  These ranges will be partitioned
# to containers configured to create automatically a user namespace.  Containers
# configured to automatically create a user namespace can still overlap with containers
# having an explicit mapping set.
# This setting is ignored when running as rootless.
# root-auto-userns-user = "storage"
#
# Auto-userns-min-size is the minimum size for a user namespace created automatically.
# auto-userns-min-size=1024
#
# Auto-userns-max-size is the minimum size for a user namespace created automatically.
# auto-userns-max-size=65536

[storage.options.overlay]
# ignore_chown_errors can be set to allow a non privileged user running with
# a single UID within a user namespace to run containers. The user can pull
# and use any image even those with multiple uids.  Note multiple UIDs will be
# squashed down to the default uid in the container.  These images will have no
# separation between the users in the container. Only supported for the overlay
# and vfs drivers.
#ignore_chown_errors = "false"

# Path to an helper program to use for mounting the file system instead of mounting it
# directly.
# MODIFIED - uncommented
mount_program = "/usr/bin/fuse-overlayfs"

# mountopt specifies comma separated list of extra mount options
mountopt = "nodev"

# Set to skip a PRIVATE bind mount on the storage home directory.
# skip_mount_home = "false"

# Size is used to set a maximum size of the container image.
# size = ""

# ForceMask specifies the permissions mask that is used for new files and
# directories.
#
# The values "shared" and "private" are accepted.
# Octal permission masks are also accepted.
#
#  "": No value specified.
#     All files/directories, get set with the permissions identified within the
#     image.
#  "private": it is equivalent to 0700.
#     All files/directories get set with 0700 permissions.  The owner has rwx
#     access to the files. No other users on the system can access the files.
#     This setting could be used with networked based homedirs.
#  "shared": it is equivalent to 0755.
#     The owner has rwx access to the files and everyone else can read, access
#     and execute them. This setting is useful for sharing containers storage
#     with other users.  For instance have a storage owned by root but shared
#     to rootless users as an additional store.
#     NOTE:  All files within the image are made readable and executable by any
#     user on the system. Even /etc/shadow within your image is now readable by
#     any user.
#
#   OCTAL: Users can experiment with other OCTAL Permissions.
#
#  Note: The force_mask Flag is an experimental feature, it could change in the
#  future.  When "force_mask" is set the original permission mask is stored in
#  the "user.containers.override_stat" xattr and the "mount_program" option must
#  be specified. Mount programs like "/usr/bin/fuse-overlayfs" present the
#  extended attribute permissions to processes within containers rather then the
#  "force_mask"  permissions.
#
# force_mask = ""

[storage.options.thinpool]
# Storage Options for thinpool

# autoextend_percent determines the amount by which pool needs to be
# grown. This is specified in terms of % of pool size. So a value of 20 means
# that when threshold is hit, pool will be grown by 20% of existing
# pool size.
# autoextend_percent = "20"

# autoextend_threshold determines the pool extension threshold in terms
# of percentage of pool size. For example, if threshold is 60, that means when
# pool is 60% full, threshold has been hit.
# autoextend_threshold = "80"

# basesize specifies the size to use when creating the base device, which
# limits the size of images and containers.
# basesize = "10G"

# blocksize specifies a custom blocksize to use for the thin pool.
# blocksize="64k"

# directlvm_device specifies a custom block storage device to use for the
# thin pool. Required if you setup devicemapper.
# directlvm_device = ""

# directlvm_device_force wipes device even if device already has a filesystem.
# directlvm_device_force = "True"

# fs specifies the filesystem type to use for the base device.
# fs="xfs"

# log_level sets the log level of devicemapper.
# 0: LogLevelSuppress 0 (Default)
# 2: LogLevelFatal
# 3: LogLevelErr
# 4: LogLevelWarn
# 5: LogLevelNotice
# 6: LogLevelInfo
# 7: LogLevelDebug
# log_level = "7"

# min_free_space specifies the min free space percent in a thin pool require for
# new device creation to succeed. Valid values are from 0% - 99%.
# Value 0% disables
# min_free_space = "10%"

# mkfsarg specifies extra mkfs arguments to be used when creating the base
# device.
# mkfsarg = ""

# metadata_size is used to set the `pvcreate --metadatasize` options when
# creating thin devices. Default is 128k
# metadata_size = ""

# Size is used to set a maximum size of the container image.
# size = ""

# use_deferred_removal marks devicemapper block device for deferred removal.
# If the thinpool is in use when the driver attempts to remove it, the driver
# tells the kernel to remove it as soon as possible. Note this does not free
# up the disk space, use deferred deletion to fully remove the thinpool.
# use_deferred_removal = "True"

# use_deferred_deletion marks thinpool device for deferred deletion.
# If the device is busy when the driver attempts to delete it, the driver
# will attempt to delete device every 30 seconds until successful.
# If the program using the driver exits, the driver will continue attempting
# to cleanup the next time the driver is used. Deferred deletion permanently
# deletes the device and all data stored in device will be lost.
# use_deferred_deletion = "True"

# xfs_nospace_max_retries specifies the maximum number of retries XFS should
# attempt to complete IO when ENOSPC (no space) error is returned by
# underlying storage device.
# xfs_nospace_max_retries = "0"
```

*docker_files/containers.conf:*
```
# The containers configuration file specifies all of the available configuration
# command-line options/flags for container engine tools like Podman & Buildah,
# but in a TOML format that can be easily modified and versioned.

# Please refer to containers.conf(5) for details of all configuration options.
# Not all container engines implement all of the options.
# All of the options have hard coded defaults and these options will override
# the built in defaults. Users can then override these options via the command
# line. Container engines will read containers.conf files in up to three
# locations in the following order:
#  1. /usr/share/containers/containers.conf
#  2. /etc/containers/containers.conf
#  3. $HOME/.config/containers/containers.conf (Rootless containers ONLY)
#  Items specified in the latter containers.conf, if they exist, override the
# previous containers.conf settings, or the default settings.

[containers]

# List of annotation. Specified as
# "key = value"
# If it is empty or commented out, no annotations will be added
#
# annotations = []

# Used to change the name of the default AppArmor profile of container engine.
#
# apparmor_profile = "container-default"

# Default way to to create a cgroup namespace for the container
# Options are:
# `private` Create private Cgroup Namespace for the container.
# `host`    Share host Cgroup Namespace with the container.
#
# cgroupns = "private"

# Control container cgroup configuration
# Determines  whether  the  container will create CGroups.
# Options are:
# `enabled`   Enable cgroup support within container
# `disabled`  Disable cgroup support, will inherit cgroups from parent
# `no-conmon` Do not create a cgroup dedicated to conmon.
#
# MODIFIED: enabled->disabled
cgroups = "disabled"

# List of default capabilities for containers. If it is empty or commented out,
# the default capabilities defined in the container engine will be added.
#
default_capabilities = [
    "CHOWN",
    "DAC_OVERRIDE",
    "FOWNER",
    "FSETID",
    "KILL",
    "NET_BIND_SERVICE",
    "SETFCAP",
    "SETGID",
    "SETPCAP",
    "SETUID",
    "SYS_CHROOT"
]

# A list of sysctls to be set in containers by default,
# specified as "name=value",
# for example:"net.ipv4.ping_group_range = 0 0".
#
# MODIFIED: Removed "net.ipv4.ping_group_range = 0 0"
# That is only needed if you want to run ping from inside the container, which we don't need to do.
default_sysctls = []

# A list of ulimits to be set in containers by default, specified as
# "<ulimit name>=<soft limit>:<hard limit>", for example:
# "nofile=1024:2048"
# See setrlimit(2) for a list of resource names.
# Any limit not specified here will be inherited from the process launching the
# container engine.
# Ulimits has limits for non privileged container engines.
#
# default_ulimits = [
#  "nofile=1280:2560",
# ]

# List of devices. Specified as
# "<device-on-host>:<device-on-container>:<permissions>", for example:
# "/dev/sdc:/dev/xvdc:rwm".
# If it is empty or commented out, only the default devices will be used
#
# devices = []

# List of default DNS options to be added to /etc/resolv.conf inside of the container.
#
# dns_options = []

# List of default DNS search domains to be added to /etc/resolv.conf inside of the container.
#
# dns_searches = []

# Set default DNS servers.
# This option can be used to override the DNS configuration passed to the
# container. The special value "none" can be specified to disable creation of
# /etc/resolv.conf in the container.
# The /etc/resolv.conf file in the image will be used without changes.
#
# dns_servers = []

# Environment variable list for the conmon process; used for passing necessary
# environment variables to conmon or the runtime.
#
# env = [
#    "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
#    "TERM=xterm",
# ]

# Pass all host environment variables into the container.
#
# env_host = false

# Default proxy environment variables passed into the container.
# The environment variables passed in include:
# http_proxy, https_proxy, ftp_proxy, no_proxy, and the upper case versions of
# these. This option is needed when host system uses a proxy but container
# should not use proxy. Proxy environment variables specified for the container
# in any other way will override the values passed from the host.
#
# http_proxy = true

# Run an init inside the container that forwards signals and reaps processes.
#
# init = false

#  Container init binary, if init=true, this is the init binary to be used for containers.
#
# init_path = "/usr/libexec/podman/catatonit"

# Default way to to create an IPC namespace (POSIX SysV IPC) for the container
# Options are:
# `private` Create private IPC Namespace for the container.
# `host`    Share host IPC Namespace with the container.
#
# ipcns = "private"

# keyring tells the container engine whether to create
# a kernel keyring for use within the container.
# keyring = true

# label tells the container engine whether to use container separation using
# MAC(SELinux) labeling or not.
# The label flag is ignored on label disabled systems.
#
# label = true

# Logging driver for the container. Available options: k8s-file and journald.
#
# log_driver = "k8s-file"

# Maximum size allowed for the container log file. Negative numbers indicate
# that no size limit is imposed. If positive, it must be >= 8192 to match or
# exceed conmon's read buffer. The file is truncated and re-opened so the
# limit is never exceeded.
#
# log_size_max = -1

# Specifies default format tag for container log messages.
# This is useful for creating a specific tag for container log messages.
# Containers logs default to truncated container ID as a tag.
#
# log_tag = ""

# Default way to to create a Network namespace for the container
# Options are:
# `private` Create private Network Namespace for the container.
# `host`    Share host Network Namespace with the container.
# `none`    Containers do not use the network
#
# netns = "private"

# Create /etc/hosts for the container.  By default, container engine manage
# /etc/hosts, automatically adding  the container's  own  IP  address.
#
# no_hosts = false

# Default way to to create a PID namespace for the container
# Options are:
# `private` Create private PID Namespace for the container.
# `host`    Share host PID Namespace with the container.
#
# pidns = "private"

# Maximum number of processes allowed in a container.
#
# pids_limit = 2048

# Indicates the networking to be used for rootless containers
# rootless_networking = "slirp4netns"

# Path to the seccomp.json profile which is used as the default seccomp profile
# for the runtime.
#
# seccomp_profile = "/usr/share/containers/seccomp.json"

# Size of /dev/shm. Specified as <number><unit>.
# Unit is optional, values:
# b (bytes), k (kilobytes), m (megabytes), or g (gigabytes).
# If the unit is omitted, the system uses bytes.
#
# shm_size = "65536k"

# Set timezone in container. Takes IANA timezones as well as "local",
# which sets the timezone in the container to match the host machine.
#
# tz = ""

# Set umask inside the container
#
# umask = "0022"

# Default way to to create a User namespace for the container
# Options are:
# `auto`        Create unique User Namespace for the container.
# `host`    Share host User Namespace with the container.
#
# userns = "host"

# Number of UIDs to allocate for the automatic container creation.
# UIDs are allocated from the "container" UIDs listed in
# /etc/subuid & /etc/subgid
#
# userns_size = 65536

# Default way to to create a UTS namespace for the container
# Options are:
# `private`        Create private UTS Namespace for the container.
# `host`    Share host UTS Namespace with the container.
#
# utsns = "private"

# List of volumes. Specified as
# "<directory-on-host>:<directory-in-container>:<options>", for example:
# "/db:/var/lib/db:ro".
# If it is empty or commented out, no volumes will be added
#
# volumes = []

# The network table contains settings pertaining to the management of
# CNI plugins.

[network]

# Path to directory where CNI plugin binaries are located.
#
# cni_plugin_dirs = ["/usr/libexec/cni"]

# The network name of the default CNI network to attach pods to.
# default_network = "podman"

# The default subnet for the default CNI network given in default_network.
# If a network with that name does not exist, a new network using that name and
# this subnet will be created.
# Must be a valid IPv4 CIDR prefix.
#default_subnet = "10.88.0.0/16"

# Path to the directory where CNI configuration files are located.
#
# network_config_dir = "/etc/cni/net.d/"

[engine]
# Index to the active service
# active_service = production

# Cgroup management implementation used for the runtime.
# Valid options "systemd" or "cgroupfs"
#
# cgroup_manager = "systemd"

# Environment variables to pass into conmon
#
# conmon_env_vars = [
#        "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
# ]

# Paths to look for the conmon container manager binary
#
# conmon_path = [
#        "/usr/libexec/podman/conmon",
#        "/usr/local/libexec/podman/conmon",
#        "/usr/local/lib/podman/conmon",
#        "/usr/bin/conmon",
#        "/usr/sbin/conmon",
#        "/usr/local/bin/conmon",
#        "/usr/local/sbin/conmon"
# ]

# Specify the keys sequence used to detach a container.
# Format is a single character [a-Z] or a comma separated sequence of
# `ctrl-<value>`, where `<value>` is one of:
# `a-z`, `@`, `^`, `[`, `\`, `]`, `^` or `_`
#
# detach_keys = "ctrl-p,ctrl-q"

# Determines whether engine will reserve ports on the host when they are
# forwarded to containers. When enabled, when ports are forwarded to containers,
# ports are held open by as long as the container is running, ensuring that
# they cannot be reused by other programs on the host. However, this can cause
# significant memory usage if a container has many ports forwarded to it.
# Disabling this can save memory.
#
# enable_port_reservation = true

# Environment variables to be used when running the container engine (e.g., Podman, Buildah).
# For example "http_proxy=internal.proxy.company.com".
# Note these environment variables will not be used within the container.
# Set the env section under [containers] table, if you want to set environment variables for the container.
# env = []

# Selects which logging mechanism to use for container engine events.
# Valid values are `journald`, `file` and `none`.
#
# MODIFIED: journald -> file
events_logger = "file"

# Path to OCI hooks directories for automatically executed hooks.
#
# hooks_dir = [
#     "/usr/share/containers/oci/hooks.d",
# ]

# Manifest Type (oci, v2s2, or v2s1) to use when pulling, pushing, building
# container images. By default image pulled and pushed match the format of the
# source image. Building/committing defaults to OCI.
# image_default_format = ""

# Default transport method for pulling and pushing for images
#
# image_default_transport = "docker://"

# Maximum number of image layers to be copied (pulled/pushed) simultaneously.
# Not setting this field, or setting it to zero, will fall back to containers/image defaults.
# image_parallel_copies = 0

# Default command to run the infra container
#
# infra_command = "/pause"

# Infra (pause) container image name for pod infra containers.  When running a
# pod, we start a `pause` process in a container to hold open the namespaces
# associated with the  pod.  This container does nothing other then sleep,
# reserving the pods resources for the lifetime of the pod.
#
# infra_image = "k8s.gcr.io/pause:3.4.1"

# Specify the locking mechanism to use; valid values are "shm" and "file".
# Change the default only if you are sure of what you are doing, in general
# "file" is useful only on platforms where cgo is not available for using the
# faster "shm" lock type.  You may need to run "podman system renumber" after
# you change the lock type.
#
# lock_type** = "shm"

# Indicates if Podman is running inside a VM via Podman Machine.
# Podman uses this value to do extra setup around networking from the
# container inside the VM to to host.
# machine_enabled = false

# MultiImageArchive - if true, the container engine allows for storing archives
# (e.g., of the docker-archive transport) with multiple images.  By default,
# Podman creates single-image archives.
#
# multi_image_archive = "false"

# Default engine namespace
# If engine is joined to a namespace, it will see only containers and pods
# that were created in the same namespace, and will create new containers and
# pods in that namespace.
# The default namespace is "", which corresponds to no namespace. When no
# namespace is set, all containers and pods are visible.
#
# namespace = ""

# Path to the slirp4netns binary
#
# network_cmd_path = ""

# Default options to pass to the slirp4netns binary.
# For example "allow_host_loopback=true"
#
# network_cmd_options = []

# Whether to use chroot instead of pivot_root in the runtime
#
# no_pivot_root = false

# Number of locks available for containers and pods.
# If this is changed, a lock renumber must be performed (e.g. with the
# 'podman system renumber' command).
#
# num_locks = 2048

# Whether to pull new image before running a container
# pull_policy = "missing"

# Indicates whether the application should be running in remote mode. This flag modifies the
# --remote option on container engines. Setting the flag to true will default
# `podman --remote=true` for access to the remote Podman service.
# remote = false

# Default OCI runtime
#
# runtime = "crun"

# List of the OCI runtimes that support --format=json.  When json is supported
# engine will use it for reporting nicer errors.
#
# runtime_supports_json = ["crun", "runc", "kata", "runsc"]

# List of the OCI runtimes that supports running containers with KVM Separation.
#
# runtime_supports_kvm = ["kata"]

# List of the OCI runtimes that supports running containers without cgroups.
#
# runtime_supports_nocgroups = ["crun"]

# Directory for persistent engine files (database, etc)
# By default, this will be configured relative to where the containers/storage
# stores containers
# Uncomment to change location from this default
#
# static_dir = "/var/lib/containers/storage/libpod"

# Number of seconds to wait for container to exit before sending kill signal.
# stop_timeout = 10

# map of service destinations
# [service_destinations]
#   [service_destinations.production]
#     URI to access the Podman service
#     Examples:
#       rootless "unix://run/user/$UID/podman/podman.sock" (Default)
#       rootfull "unix://run/podman/podman.sock (Default)
#       remote rootless ssh://engineering.lab.company.com/run/user/1000/podman/podman.sock
#       remote rootfull ssh://root@10.10.1.136:22/run/podman/podman.sock
#     uri = "ssh://user@production.example.com/run/user/1001/podman/podman.sock"
#     Path to file containing ssh identity key
#     identity = "~/.ssh/id_rsa"

# Directory for temporary files. Must be tmpfs (wiped after reboot)
#
# tmp_dir = "/run/libpod"

# Directory for libpod named volumes.
# By default, this will be configured relative to where containers/storage
# stores containers.
# Uncomment to change location from this default.
#
# volume_path = "/var/lib/containers/storage/volumes"

# Paths to look for a valid OCI runtime (crun, runc, kata, runsc, etc)
[engine.runtimes]
# crun = [
#            "/usr/bin/crun",
#            "/usr/sbin/crun",
#            "/usr/local/bin/crun",
#            "/usr/local/sbin/crun",
#            "/sbin/crun",
#            "/bin/crun",
#            "/run/current-system/sw/bin/crun",
# ]

# kata = [
#            "/usr/bin/kata-runtime",
#            "/usr/sbin/kata-runtime",
#            "/usr/local/bin/kata-runtime",
#            "/usr/local/sbin/kata-runtime",
#            "/sbin/kata-runtime",
#            "/bin/kata-runtime",
#            "/usr/bin/kata-qemu",
#            "/usr/bin/kata-fc",
# ]

# runc = [
#        "/usr/bin/runc",
#        "/usr/sbin/runc",
#        "/usr/local/bin/runc",
#        "/usr/local/sbin/runc",
#        "/sbin/runc",
#        "/bin/runc",
#        "/usr/lib/cri-o-runc/sbin/runc",
# ]

# runsc = [
#            "/usr/bin/runsc",
#            "/usr/sbin/runsc",
#            "/usr/local/bin/runsc",
#            "/usr/local/sbin/runsc",
#            "/bin/runsc",
#            "/sbin/runsc",
#            "/run/current-system/sw/bin/runsc",
# ]

[engine.volume_plugins]
# testplugin = "/run/podman/plugins/test.sock"

# The [engine.volume_plugins] table MUST be the last entry in this file.
# (Unless another table is added)
# TOML does not provide a way to end a table other than a further table being
# defined, so every key hereafter will be part of [volume_plugins] and not the
# main config.

[secret]
# driver = "file"

[secret.opts]
# root = "/example/directory"

```

2. Start a shell inside a container that has podman: `docker run -it --rm --entrypoint /bin/bash --privileged podman_perf_test`
3.
```
root@a7a8887add26:/# mkdir perf_test
root@a7a8887add26:/# cd perf_test
root@a7a8887add26:/perf_test# cat - > Dockerfile

# paste in the following sample dockerfile:

# x or y
ARG app_type

FROM python:3.8-slim-buster AS common

RUN mkdir /dir1
RUN mkdir /dir2
RUN mkdir /dir3

FROM common AS x

RUN mkdir /x

FROM common AS y

RUN mkdir /y

FROM ${app_type} AS final

ARG app_type

RUN touch /${app_type}/final.txt


# ctrl+d, back to the shell


root@a7a8887add26:/perf_test# podman build . --no-cache -t perf_test --build-arg app_type=x --target final
```

**Describe the results you received:**
Build took about 22 seconds after the base image was pulled. Each mkdir took multiple seconds.

**Describe the results you expected:**
I expect a build like this to take only a few seconds after the base image is pulled and each mkdir to take less than a second.

**Output of `rpm -q buildah` or `apt list buildah`:**

```
root@a7a8887add26:/perf_test# apt list buildah
Listing... Done
root@a7a8887add26:/perf_test# apt list podman
Listing... Done
podman/now 100:3.2.3-1 amd64 [installed,local]
```

**Output of `buildah version`:**

(not using buildah directly, using it through podman)

**Output of `podman version` if reporting a `podman build` issue:**

```
root@a7a8887add26:/perf_test# podman version
Version:      3.2.3
API Version:  3.2.3
Go Version:   go1.15.9
Built:        Thu Jan  1 00:00:00 1970
OS/Arch:      linux/amd64
```

**Output of `cat /etc/*release`:**

```
root@a7a8887add26:/perf_test# cat /etc/*release
PRETTY_NAME="Debian GNU/Linux 10 (buster)"
NAME="Debian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```

**Output of `uname -a`:**

```
root@a7a8887add26:/perf_test# uname -a
Linux a7a8887add26 5.10.25-linuxkit #1 SMP Tue Mar 23 09:27:39 UTC 2021 x86_64 GNU/Linux
```

**Output of `cat /etc/containers/storage.conf`:**

(same as the storage.conf above)

**Output of `podman info`:**

```
root@a7a8887add26:/perf_test# podman info
host:
  arch: amd64
  buildahVersion: 1.21.3
  cgroupControllers:
  - cpuset
  - cpu
  - cpuacct
  - blkio
  - memory
  - devices
  - freezer
  - net_cls
  - perf_event
  - net_prio
  - hugetlb
  - pids
  - rdma
  cgroupManager: cgroupfs
  cgroupVersion: v1
  conmon:
    package: 'conmon: /usr/libexec/podman/conmon'
    path: /usr/libexec/podman/conmon
    version: 'conmon version 2.0.27, commit: '
  cpus: 2
  distribution:
    distribution: debian
    version: "10"
  eventLogger: file
  hostname: a7a8887add26
  idMappings:
    gidmap: null
    uidmap: null
  kernel: 5.10.25-linuxkit
  linkmode: dynamic
  memFree: 1256357888
  memTotal: 3137761280
  ociRuntime:
    name: crun
    package: 'crun: /usr/bin/crun'
    path: /usr/bin/crun
    version: |-
      crun version 0.20.1.5-925d-dirty
      commit: 0d42f1109fd73548f44b01b3e84d04a279e99d2e
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +YAJL
  os: linux
  remoteSocket:
    path: /run/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: false
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: false
  serviceIsRemote: false
  slirp4netns:
    executable: ""
    package: ""
    version: ""
  swapFree: 275017728
  swapTotal: 536866816
  uptime: 7h 43m 50s (Approximately 0.29 days)
registries:
  search:
  - docker.io
  - quay.io
store:
  configFile: /etc/containers/storage.conf
  containerStore:
    number: 0
    paused: 0
    running: 0
    stopped: 0
  graphDriverName: overlay
  graphOptions:
    overlay.mount_program:
      Executable: /usr/bin/fuse-overlayfs
      Package: 'fuse-overlayfs: /usr/bin/fuse-overlayfs'
      Version: |-
        fusermount3 version: 3.4.1
        fuse-overlayfs: version 1.5
        FUSE library version 3.4.1
        using FUSE kernel interface version 7.27
    overlay.mountopt: nodev
  graphRoot: /var/lib/containers/storage
  graphStatus:
    Backing Filesystem: overlayfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "false"
  imageStore:
    number: 85
  runRoot: /run/containers/storage
  volumePath: /var/lib/containers/storage/volumes
version:
  APIVersion: 3.2.3
  Built: 0
  BuiltTime: Thu Jan  1 00:00:00 1970
  GitCommit: ""
  GoVersion: go1.15.9
  OsArch: linux/amd64
  Version: 3.2.3
```

I also tried running the `podman build` with `--squash` and separately with `--layers=false` In both cases the first two mkdir's in the first build stage are fast, with the final one being slow, which would largely solve the performance problem for us. However, when running interactively this results in a prompt:

```
root@a7a8887add26:/perf_test# podman build . --no-cache -t perf_test --build-arg app_type=x --target final --squash
STEP 1: FROM python:3.8-slim-buster AS common
STEP 2: RUN mkdir /dir1
STEP 3: RUN mkdir /dir2
STEP 4: RUN mkdir /dir3
--> 2781c7a7fb6
STEP 5: FROM 2781c7a7fb6233baac2202922c5e8021c31de73916b07ed5f2fced33425b3190 AS x
STEP 6: RUN mkdir /x
STEP 7: FROM 2781c7a7fb6233baac2202922c5e8021c31de73916b07ed5f2fced33425b3190 AS y
STEP 8: RUN mkdir /y
STEP 9: FROM x AS final
? Please select an image:
  ▸ docker.io/library/x:latest
    quay.io/x:latest
```

When running non-interactively, it looks for an image called "x:latest" in docker.io/library and quay.io. I expected it to use the "x" build stage like it does when not building with` --squash` or `--layers=false`. Is that a bug or is it expected behavior when using those flags?

**Comment by rhatdan** [comment]
When  you do this rootfull, does it perform better?

**Comment by LHCGreg** [comment]
> When you do this rootfull, does it perform better?

It is already being run as root. I have not been able to get rootless podman in a container working.

**Comment by flouthoc** [comment]
Hi @LHCGreg , I tried reproducing this, current buildah only takes about `9 seconds` as compared to your original reproducer which was reported taking `22+` seconds. I'd request you to try this on latest buildah inside a container. I think this issue is does not persists anymore therefore I am closing this issue but please feel free to re-open if you think issue still persists.

Here is invoke which i used for launching buildah inside a container
```console
sudo podman run -it --device /dev/fuse:rw --security-opt seccomp=unconfined --security-opt apparmor=unconfined quay.io/buildah/stable:latest bash
```

**Comment by flouthoc** [comment]
@LHCGreg Just to try out it its working fine on buildah could you replace `FROM ${app_type} AS final` in dockerfile as `FROM x AS final` if it works fine I mean the original slow issue, we can open a new issue to parse build args early and that might be a different issue.

**Comment by LHCGreg** [comment]
@flouthoc I confirmed that when replacing `FROM ${app_type} AS final` with `FROM x AS final`, buildah does not have the performance problem described in this issue and the build completes successfully. I can create a new issue for supporting dynamic FROM when using `buildah bud`.

For podman, I tried using quay.io/podman/stable:latest instead of my custom image and still got slow performance when using `podman build`. Should I open an issue for this on the podman repo instead of here then?

I also tried installing buildah inside my custom image and got the same results as with quay.io/buildah/stable:latest - good performance but dynamic FROM doesn't work.

Summary:

- custom image (dockerfile in comment earlier today)
- - podman 3.4.2: slow performance, dynamic FROM ok
- - buildah 1.21.3: fast performance, dynamic FROM not ok
- quay.io/podman/stable:latest:
- - podman 3.4.4: slow performance, dynamic FROM ok
- quay.io/buildah/stable:latest:
- - buildah 1.23.3: fast performance, dynamic FROM not ok

**Comment by flouthoc** [comment]
Okay i see the problem here but its also not a problem `build` is slow with podman because it uses `--layers` by default so when `--layers` is used `buildah` creates `intermediate images` which can be used as cache later for future builds, here buildah takes time because it commits these  `intermediate images`.

So if you do `podman build --layers=false` it should give good performance since we are not committing `intermediate images` so no time wasted there.

However `dynamic FROM` is still a legit issue hence i'm creating a new issue for it.

**Comment by LHCGreg** [comment]
Thanks for the insight into why `podman build` and `buildah bud` were behaving differently. If #3939 is fixed or if we refactor to use a preprocessor to pre-generate multiple dockerfiles instead of using dynamic build stages with a single dockerfile, that should solve our performance problem by allowing us to use --layers=false. Our image build is happening inside a container with no persistent storage that gets terminated when the image build is complete, so caching intermediate layers is not useful in our case.

However, to circle back to the original issue, the performance of saving each cached intermediate layer when `--layers=true` seems slower than it should be. It shouldn't take multiple seconds to save a layer consisting of a single empty directory, right? Perhaps this issue should be reopened?

**Comment by rhatdan** [comment]
Any chance you are using the VFS driver?

**Comment by flouthoc** [comment]
@LHCGreg PR https://github.com/containers/buildah/pull/3947 should fix dynamic FROM issue and should be out in next release or podman `4.1` if gets merged early and should unblock you.

For second part I think we can benchmark `commit` and check if there is any room for optimization and re-open issue if i find any abnormal performance issues there.

**Comment by LHCGreg** [comment]
@rhatdan : As per `podman info` in the original post, I am not using the VFS driver.

@flouthoc : Thanks! One last question: When that fix is released, how would one install that new version in Debian stable? We have been getting podman from https://download.opensuse.org/repositories/devel:/kubic:/libcontainers:/stable/Debian_11/ for fresher versions than in the main Debian repository but that now seems to be a few versions behind, even in the 3.x line. Current version there is podman 3.4.2 while the latest 3.x release seems to be 3.4.7. https://podman.io/getting-started/installation only mentions the main Debian repository which is even older, 3.0.1. Is building from source the only option?

**Comment by flouthoc** [comment]
@LHCGreg Most likely in the next release of buildah and `4.1` of podman (if this gets merged today) https://github.com/containers/buildah/pull/3947. I'll notify you whenever this PR is out.

> Debian repository but that now seems to be a few versions behind, even in the 3.x line. Current version there is podman 3.4.2 while the latest 3.x release seems to be 3.4.7. https://podman.io/getting-started/installation only mentions the main Debian repository which is even older, 3.0.1. Is building from source the only option?

Ah i see debian release is running few versions behind but no issues I think building from source is not so difficult https://podman.io/getting-started/installation#building-from-scratch . You can always open issues/discussions on `buildah/podman` repo if you need any help with this.
